compiler/damlc/daml-doc/src/DA/Daml/Doc/HaddockParse.hs

```diff
@@ -422,84 +455,110 @@ docToText = DocText . T.strip . T.unlines . go . T.lines . T.pack . unpackHDS
+tyConAnchor :: DocCtx -> TyCon -> Maybe Anchor
+tyConAnchor DocCtx{..} tycon = do
+    let ghcName = tyConName tycon
+        name = Typename . packName $ ghcName
+        mod = maybe dc_mod packModule (nameModule_maybe ghcName)
+        anchorFn
+            | isClassTyCon tycon = classAnchor
+            | isDataTyCon tycon = dataAnchor
+            | otherwise = typeAnchor
+    guard (not (isWiredInName ghcName))
+    Just (anchorFn mod name)
 
-  HsOpTy _ ty1 (L _ op) ty2 ->
-    TypeApp Nothing (Typename $ toText op) [ hsTypeToType ty1, hsTypeToType ty2 ]
+---------------------------------------------------------------------
 
-  HsParTy _ ty  -> hsTypeToType ty
+-- | Extract context from GHC type. Returns Nothing if there are no constraints.
+typeToContext :: DocCtx -> TyCoRep.Type -> Maybe DDoc.Type
+typeToContext dc ty =
+    let ctx = typeToConstraints dc ty
+    in guard (notNull ctx) >> Just (TypeTuple ctx)
+
+-- | Extract constraints from GHC type, returning list of constraints.
+typeToConstraints :: DocCtx -> TyCoRep.Type -> [DDoc.Type]
+typeToConstraints dc = \case
+    FunTy a@(TyConApp tycon _) b | isClassTyCon tycon ->
+        typeToType dc a : typeToConstraints dc b
+    FunTy _ b ->
+        typeToConstraints dc b
+    ForAllTy _ b -> -- TODO: I think forall can introduce constraints?
+        typeToConstraints dc b
+    _ -> []
+
+
+-- | Convert GHC Type into a damldoc type, ignoring constraints.
+typeToType :: DocCtx -> TyCoRep.Type -> DDoc.Type
+typeToType ctx = \case
+    TyVarTy var -> TypeApp Nothing (Typename $ packId var) []
+    AppTy a b ->
+        case typeToType ctx a of
+            TypeApp m f bs -> TypeApp m f (bs <> [typeToType ctx b]) -- flatten app chains
+            TypeFun _ -> unexpected "function type in a type app"
+            TypeList _ -> unexpected "list type in a type app"
+            TypeTuple _ -> unexpected "tuple type in a type app"
+    TyConApp tycon bs ->
+        typeApp
+            (tyConAnchor ctx tycon)
+            (Typename . packName . tyConName $ tycon)
+            (map (typeToType ctx) bs)
+
+    -- ignore context
+    ForAllTy _ b -> typeToType ctx b
+    FunTy (TyConApp tycon _) b | isClassTyCon tycon ->
+        typeToType ctx b
+
+    FunTy a b ->
+        case typeToType ctx b of
+            TypeFun bs -> TypeFun (typeToType ctx a : bs) -- flatten function types
+            b' -> TypeFun [typeToType ctx a, b']
+
+    CastTy a _ -> typeToType ctx a
+    LitTy x -> TypeApp Nothing (Typename $ toText x) []
+    CoercionTy _ -> unexpected "coercion" -- TODO?
 
-  XHsType _t -> unexpected "XHsType"
+  where
+    -- | Deal with list and tuple type constructors specially.
+    typeApp _ (Typename "[]") [x] = TypeList x
+    typeApp _ (Typename c) xs | isTupleCons (T.unpack c) = TypeTuple xs
+    typeApp anchor name xs = TypeApp anchor name xs
+
+    -- | Is the input a tuple constructor?
+    isTupleCons xs = and
```

Comment:
I’d be more comfortable if we could pass the `TyCon` into `typeApp` and then use GHC’s `isTupleTyCon`.

Reply:
good idea :)